From d199f70e8826b669252d646237eeff759daaa6e2 Mon Sep 17 00:00:00 2001
From: Artem Lapkin <art@khadas.com>
Date: Fri, 18 Jun 2021 13:23:53 +0800
Subject: [PATCH] PCI: dwc: meson add quirk

Device set same 256 bytes maximum read request size equal MAX_READ_REQ_SIZE
was find some issue with HDMI scrambled picture and nvme devices
at intensive writing...

[    4.798971] nvme 0000:01:00.0: fix MRRS from 512 to 256

This quirk setup same MRRS if we try solve this problem with pci=pcie_bus_perf

Signed-off-by: Artem Lapkin <art@khadas.com>
---
 drivers/pci/controller/dwc/pci-meson.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/pci/controller/dwc/pci-meson.c b/drivers/pci/controller/dwc/pci-meson.c
index a2656672c..dd7ae3ffd 100644
--- a/drivers/pci/controller/dwc/pci-meson.c
+++ b/drivers/pci/controller/dwc/pci-meson.c
@@ -463,6 +463,32 @@ static int meson_pcie_probe(struct platform_device *pdev)
 	return 0;
 }
 
+static void meson_pcie_quirk(struct pci_dev *dev)
+{
+    int mrrs;
+
+    /* no need quirk */
+    if (pcie_bus_config != PCIE_BUS_DEFAULT)
+	return;
+
+    /* no need for root bus */
+    if (pci_is_root_bus(dev->bus))
+	return;
+
+    mrrs = pcie_get_readrq(dev);
+
+    /*
+     * set same 256 bytes maximum read request size equal MAX_READ_REQ_SIZE
+     * was find some issue with HDMI scrambled picture and nvme devices
+     * at intensive writing...
+     */
+    if (mrrs != MAX_READ_REQ_SIZE ) {
+	dev_info(&dev->dev, "fix MRRS from %d to %d\n", mrrs, MAX_READ_REQ_SIZE);
+	pcie_set_readrq(dev, MAX_READ_REQ_SIZE);
+    }
+}
+DECLARE_PCI_FIXUP_ENABLE(PCI_ANY_ID, PCI_ANY_ID, meson_pcie_quirk);
+
 static const struct of_device_id meson_pcie_of_match[] = {
 	{
 		.compatible = "amlogic,axg-pcie",
-- 
2.25.1

